### PR TITLE
[Dynamic Instrumentation] Add Snapshot Exploration Test infrastructure

### PR DESCRIPTION
**Pull request by dudikeleti**
## Summary of changes

- Introduce a new internal **snapshot exploration test** mode for Dynamic Instrumentation that generates a JSON probe file for a third-party exploration test project, runs those probes in the test host, and writes per-process CSV reports for captured snapshots and probe statuses.
- Add Nuke targets `SetUpSnapshotExplorationTests` / `RunSnapshotExplorationTests` to generate `SnapshotExplorationTestProbes.json`, run the selected debugger exploration test, and analyze the generated reports/logs after the test completes.
- Add managed runtime plumbing that is activated only inside a vstest host when `DD_INTERNAL_SNAPSHOT_EXPLORATION_TEST_ROOT_PATH` is present: `SnapshotExplorationTestSink`, `SnapshotExplorationProbeStatusPoller` + `SnapshotExplorationProbeStatusReporter`, `SnapshotExplorationConstants`, `NoOpDebuggerUploader`, and a snapshot-exploration init path on `DebuggerManager`.
- Extract an `ISnapshotSink` interface so normal debugger uploads continue to use `SnapshotSink`, while snapshot exploration can use `SnapshotExplorationTestSink` to write report rows instead of buffering uploads for an agent.
- Treat `DD_INTERNAL_SNAPSHOT_EXPLORATION_TEST_ROOT_PATH` as an internal test-harness environment variable, not a generated `ConfigurationKeys.Debugger` setting, and exclude it from configuration telemetry registration validation.
- Add unit tests for `SnapshotExplorationTestSink` report writing, CSV escaping, duplicate probe handling, and invalid/partial snapshot grading.

## Reason for change

- We need a mechanism to validate Dynamic Instrumentation across many real-world methods without standing up an agent or live RCM flow.
- The exploration-test harness gives us broad third-party method coverage, while the new sink/reporting path gives a build-side artifact that identifies which generated probes produced valid snapshots and which failed or were skipped.

## Implementation details

- Snapshot exploration is enabled only when `DD_INTERNAL_SNAPSHOT_EXPLORATION_TEST_ROOT_PATH` is set and the current process is detected as a vstest host. Child processes may inherit the env var, but they keep the normal debugger path.
- The Nuke setup target reflects over built exploration-test assemblies and writes `SnapshotExplorationTestProbes.json` with `LOG_PROBE` definitions, `captureSnapshot=true`, and a deterministic sampled `when` expression for some instance methods.
- Exploration runs are agent-less: tracing, telemetry retries, and agent discovery are disabled through the test environment so report/log analysis focuses on debugger behavior.
- In snapshot exploration mode, snapshot/log uploaders share the same report-writing sink, diagnostics and symbol upload are replaced with no-op uploaders, and probe status polling writes local CSV status reports.
- This PR only adds the local/Nuke infrastructure; it does **not** add or modify Azure Pipelines stages, schedules, or CI gating for snapshot exploration runs.

## Test coverage

- `SnapshotExplorationTestSinkTests` (7 cases: report content, nested-logger handling, CSV escaping, probe-id dedup, and invalid/partial-snapshot grading).
- The end-to-end `RunSnapshotExplorationTests` Nuke target is intentionally **not** wired into CI in this PR because it is long-running and debugger-specific.

## Other details

- Replaces draft PR #6064 (the original branch contained both this feature and several unrelated bug fixes; this PR is the feature only).
- Follow-up work: add a dedicated CI stage/job that runs snapshot exploration only for debugger-related changes and on a daily master schedule.

### CODE REVIEW

File: tracer/src/Datadog.Trace/Debugger/DebuggerSettings.cs [hunk 1]
```diff
@@ -151,9 +152,16 @@ public DebuggerSettings(IConfigurationSource? source, IConfigurationTelemetry te
...
+
             SymbolDatabaseCompressionEnabled = config.WithKeys(ConfigurationKeys.Debugger.SymbolDatabaseCompressionEnabled).AsBool(true);
 
-            ProbeFile = config.WithKeys(ConfigurationKeys.Debugger.DynamicInstrumentationProbeFile).AsString() ?? string.Empty;
+            SnapshotExplorationTestRootPath = config.WithKeys(ConfigurationKeys.Debugger.SnapshotExplorationTestRootPath).AsString(string.Empty);
+            // Snapshot exploration is an internal test mode for the test host only. Child processes may inherit the
+            // root path, but they should keep the normal debugger sinks/pollers/rate limits.
+            IsSnapshotExplorationTestEnabled =
+                !StringUtil.IsNullOrEmpty(SnapshotExplorationTestRootPath) &&
+                SnapshotExplorationConstants.IsRunningInTestHost();
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/DebuggerSettings.cs]: decription updated

File: tracer/src/Datadog.Trace/Debugger/Sink/ISnapshotSink.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+using System;
+using System.Collections.Generic;
+
+namespace Datadog.Trace.Debugger.Sink;
+
+internal interface ISnapshotSink : IDisposable
+{
+    public void Add(string probeId, string? snapshot);
+
+    IList<string> GetSnapshots();
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Sink/ISnapshotSink.cs:16]: 83651dd35287a77ba7991323be1920acd8033ddc

File: tracer/src/Datadog.Trace/Debugger/DebuggerManager.SnapshotExploration.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                Thread.Sleep(SnapshotExplorationConstants.NativeLogPollInterval);
+            }
+
+            Log.Warning("Timeout waiting for native to receive probes. Proceeding anyway.");
+        }
+
+        private static bool NativeLogContainsProbeReceivedMarker(string logDir)
+        {
+            foreach (var logFile in Directory.GetFiles(logDir, SnapshotExplorationConstants.NativeLogFilePattern))
+            {
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/DebuggerManager.SnapshotExploration.cs:69]: b65d7ef2613639abaedb1ef9177cd1ca8e0ea39a

File: tracer/src/Datadog.Trace/Debugger/Sink/SnapshotExplorationTestSink.cs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+            private bool _disposed;
+
+            public ProbeReportWriter(string folderPath)
+            {
+                if (folderPath == null)
+                {
+                    throw new ArgumentNullException(nameof(folderPath));
+                }
+
+#if NET6_0_OR_GREATER
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Sink/SnapshotExplorationTestSink.cs:73]: fdecd4c538ffeca40ddec40573a14933bc5ffa74

File: tracer/build/_build/Build.SnapshotExplorationTest.cs [hunk 5]
```diff
@@ -0,0 +1,1565 @@
...
+    ProbeReportInfo ParseCsvLine(string line)
+    {
+        var parts = line.Split(',');
+
+        if (parts.Length != 4 || parts.Any(string.IsNullOrWhiteSpace))
+        {
+            var probeId = parts.Length > 0 ? parts[0].Trim() : "missing id";
+            var type = parts.Length > 1 ? parts[1].Trim() : "missing type";
+            var method = parts.Length > 1 ? parts[2].Trim() : "missing method";
+            return new ProbeReportInfo(probeId, type + "." + method, false, true);
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/build/_build/Build.SnapshotExplorationTest.cs:1074]: d84355d21c2f2822d15cfeffb249186f3dd504db

File: tracer/build/_build/Build.SnapshotExplorationTest.cs [hunk 6]
```diff
@@ -0,0 +1,1565 @@
...
+            {
+                continue;
+            }
+
+            reportInfo.AddRange(
+            File.ReadLines(file)
+                .Skip(1) // Skip the header row
+                .Select(ParseCsvLine)
+                .ToList());
+        }
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/build/_build/Build.SnapshotExplorationTest.cs]: 8a2bfddc6bf5a576ffae6864b394810e82497461

File: tracer/build/_build/Build.SnapshotExplorationTest.cs [hunk 7]
```diff
@@ -0,0 +1,1565 @@
...
+        FileSystemTasks.EnsureCleanDirectory(Path.Combine(BuildDataDirectory, "logs"));
+
+        var frameworks = Framework == null ? testDescription.SupportedFrameworks : new[] { Framework };
+        foreach (var framework in frameworks)
+        {
+            if (!testDescription.IsFrameworkSupported(framework))
+            {
+                Logger.Information($"Skipping exploration test: {testDescription.Name}.");
+                Logger.Warning($"The framework '{framework}' is not listed in the project's target frameworks of {testDescription.Name}");
+            }
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/build/_build/Build.SnapshotExplorationTest.cs:87]: 4c742c7dba52dd0ab3f384a9f30bbb262105f857

File: tracer/build/_build/Build.SnapshotExplorationTest.cs [hunk 8]
```diff
@@ -0,0 +1,1558 @@
...
+                 || assembly.ManifestModule.IsResource()
+                 || IgnoredNamespaces.Any(name => assembly.ManifestModule.Name.ToLower().StartsWith(name)))
+                {
+                    continue;
+                }
+
+                var symbolExtractor = createMethod?.Invoke(null, new object[] { assembly });
+                if (getClassSymbols?.Invoke(symbolExtractor, null) is not IEnumerable classSymbols)
+                {
+                    continue;
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/build/_build/Build.SnapshotExplorationTest.cs:166]: eb0b9266521a2cd29e2fd7ac0b3904fe3e57c1b2

File: tracer/src/Datadog.Trace/Debugger/DebuggerSettings.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/DebuggerSettings.cs]: fixed

File: tracer/src/Datadog.Trace/Generated/net461/Datadog.Trace.SourceGenerators/ConfigurationKeysGenerator/ConfigurationKeys.Debugger.g.cs [hunk 9]
```diff
@@ -137,6 +137,12 @@ internal static class Debugger
         /// </summary>
         public const string InternalDynamicInstrumentationMaxProbesPerType = "DD_INTERNAL_DYNAMIC_INSTRUMENTATION_MAX_PROBES_PER_TYPE";
 
+        /// <summary>
+        /// Internal configuration key for the snapshot exploration test root path.
+        /// The tracer derives the probes CSV path and report output folder from this root.
+        /// </summary>
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Generated/net461/Datadog.Trace.SourceGenerators/ConfigurationKeysGenerator/ConfigurationKeys.Debugger.g.cs]: 269184408b46295cb5dd9ca547e7b7124d54407b

File: tracer/src/Datadog.Trace/Generated/netcoreapp3.1/Datadog.Trace.SourceGenerators/ConfigurationKeysGenerator/ConfigurationKeys.Debugger.g.cs [hunk 10]
```diff
@@ -137,6 +137,12 @@ internal static class Debugger
         /// </summary>
         public const string InternalDynamicInstrumentationMaxProbesPerType = "DD_INTERNAL_DYNAMIC_INSTRUMENTATION_MAX_PROBES_PER_TYPE";
 
+        /// <summary>
+        /// Internal configuration key for the snapshot exploration test root path.
+        /// The tracer derives the probes CSV path and report output folder from this root.
+        /// </summary>
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Generated/netcoreapp3.1/Datadog.Trace.SourceGenerators/ConfigurationKeysGenerator/ConfigurationKeys.Debugger.g.cs]: 269184408b46295cb5dd9ca547e7b7124d54407b

File: tracer/src/Datadog.Trace/Generated/netstandard2.0/Datadog.Trace.SourceGenerators/ConfigurationKeysGenerator/ConfigurationKeys.Debugger.g.cs [hunk 11]
```diff
@@ -137,6 +137,12 @@ internal static class Debugger
         /// </summary>
         public const string InternalDynamicInstrumentationMaxProbesPerType = "DD_INTERNAL_DYNAMIC_INSTRUMENTATION_MAX_PROBES_PER_TYPE";
 
+        /// <summary>
+        /// Internal configuration key for the snapshot exploration test root path.
+        /// The tracer derives the probes CSV path and report output folder from this root.
+        /// </summary>
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Generated/netstandard2.0/Datadog.Trace.SourceGenerators/ConfigurationKeysGenerator/ConfigurationKeys.Debugger.g.cs]: 269184408b46295cb5dd9ca547e7b7124d54407b

File: tracer/src/Datadog.Trace/Generated/net6.0/Datadog.Trace.SourceGenerators/ConfigurationKeysGenerator/ConfigurationKeys.Debugger.g.cs [hunk 12]
```diff
@@ -137,6 +137,12 @@ internal static class Debugger
         /// </summary>
         public const string InternalDynamicInstrumentationMaxProbesPerType = "DD_INTERNAL_DYNAMIC_INSTRUMENTATION_MAX_PROBES_PER_TYPE";
 
+        /// <summary>
+        /// Internal configuration key for the snapshot exploration test root path.
+        /// The tracer derives the probes CSV path and report output folder from this root.
+        /// </summary>
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Generated/net6.0/Datadog.Trace.SourceGenerators/ConfigurationKeysGenerator/ConfigurationKeys.Debugger.g.cs]: 269184408b46295cb5dd9ca547e7b7124d54407b

File: tracer/src/Datadog.Trace/Configuration/supported-configurations.yaml [hunk 13]
```diff
@@ -1177,6 +1177,15 @@ supportedConfigurations:
...
       Default value is <c>0</c>, no probe-count limit is enforced.
+  DD_INTERNAL_SNAPSHOT_EXPLORATION_TEST_ROOT_PATH:
+  - implementation: A
+    type: string
+    default: null
+    product: Debugger
+    const_name: SnapshotExplorationTestRootPath
+    documentation: |-
+      Internal configuration key for the snapshot exploration test root path.
+      The tracer derives the probes CSV path and report output folder from this root.
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Configuration/supported-configurations.yaml]: 269184408b46295cb5dd9ca547e7b7124d54407b

File: tracer/build/_build/Build.SnapshotExplorationTest.cs [hunk 14]
```diff
@@ -0,0 +1,1563 @@
...
+        {
+            var probeId = parts.Length > 0 ? parts[0].Trim() : "missing id";
+            var type = parts.Length > 1 ? parts[1].Trim() : "missing type";
+            var method = parts.Length > 2 ? parts[2].Trim() : "missing method";
+            return new ProbeReportInfo(probeId, type + "." + method, false, true);
+        }
+
+        var isValid = bool.TryParse(parts[3].Trim(), out var parsedIsValid) && parsedIsValid;
+        return new ProbeReportInfo(parts[0].Trim(), parts[1].Trim() + "." + parts[2].Trim(), isValid, false);
+    }
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/build/_build/Build.SnapshotExplorationTest.cs:1115]: 269184408b46295cb5dd9ca547e7b7124d54407b

File: tracer/src/Datadog.Trace/Debugger/Sink/SnapshotExplorationTestSink.cs [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+                var fileName = Environment.ProcessId + "_" + _fileName;
+#else
+                var fileName = Process.GetCurrentProcess().Id + "_" + _fileName;
+#endif
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Sink/SnapshotExplorationTestSink.cs]: e248accf115fbf2d140f16c97fa55278583485eb

File: tracer/build/_build/Build.SnapshotExplorationTest.cs [hunk 16]
```diff
@@ -0,0 +1,1606 @@
...
+        // Rolled files use pattern: dotnet-tracer-managed-{process}-{pid}_{sequence}.log
+        var rolledNative = nativeLogs.Where(f => System.Text.RegularExpressions.Regex.IsMatch(Path.GetFileName(f), @"_\d+\.log$")).ToList();
+        var rolledManaged = managedLogs.Where(f => System.Text.RegularExpressions.Regex.IsMatch(Path.GetFileName(f), @"_\d+\.log$")).ToList();
+
+        if (rolledNative.Count > 1 || rolledManaged.Count > 1)
+        {
+            // We read ALL rolled files, but warn about the volume for awareness
+            return $"Log rolling occurred: {rolledNative.Count} native log files, {rolledManaged.Count} managed log files. " +
+                   $"All files are being analyzed, but analysis may be slow with {(totalNativeSize + totalManagedSize) / 1_000_000}MB of logs.";
+        }
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/build/_build/Build.SnapshotExplorationTest.cs]: e248accf115fbf2d140f16c97fa55278583485eb

File: tracer/build/_build/Build.SnapshotExplorationTest.cs [hunk 17]
```diff
@@ -0,0 +1,1606 @@
...
+        // Check for potential log rolling that could cause incomplete data
+        var logRollingWarning = CheckForLogRolling();
+
+        stepWatch.Restart();
+        var probesReport = ReadReportedSnapshotProbesIds(reportFolderPath);
+        timings["ReadProbesReport"] = stepWatch.Elapsed;
+        if (probesReport == null || definedProbes.Count == 0)
+        {
+            throw new Exception("Snapshot exploration test failed. Could not read report file");
+        }
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/build/_build/Build.SnapshotExplorationTest.cs]: e248accf115fbf2d140f16c97fa55278583485eb

File: tracer/build/_build/Build.SnapshotExplorationTest.cs [hunk 18]
```diff
@@ -0,0 +1,1606 @@
...
+
+        stepWatch.Restart();
+        var installedProbeIds = ReadInstalledProbeIdsFromProbeStatusReport(reportFolderPath, definedProbes, out var probeStatusReportBytes);
+        timings["ReadProbeStatusReport(installed)"] = stepWatch.Elapsed;
+        fileSizes["ProbeStatusReport"] = probeStatusReportBytes;
+        if (installedProbeIds == null || definedProbes.Count == 0)
+        {
+            throw new Exception("Snapshot exploration test failed. Could not read probe status report");
+        }
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/build/_build/Build.SnapshotExplorationTest.cs]: e248accf115fbf2d140f16c97fa55278583485eb

File: tracer/src/Datadog.Trace/Debugger/DebuggerSettings.cs [hunk 19]
```diff
@@ -202,6 +213,15 @@ public DebuggerSettings(IConfigurationSource? source, IConfigurationTelemetry te
...
         public string ProbeFile { get; }
 
+        public bool IsSnapshotExplorationTestEnabled { get; }
+
+        public string SnapshotExplorationTestRootPath { get; }
+
+        public string SnapshotExplorationTestReportFolderPath =>
+            StringUtil.IsNullOrEmpty(SnapshotExplorationTestRootPath)
+                ? string.Empty
+                : Path.Combine(SnapshotExplorationTestRootPath, "SnapshotExplorationTestReport");
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/DebuggerSettings.cs:223]: 80603ff517879032bcef0507c34f20f2fafea7cc

File: tracer/src/Datadog.Trace/Debugger/DebuggerSettings.cs [hunk 20]
```diff
@@ -151,9 +152,19 @@ public DebuggerSettings(IConfigurationSource? source, IConfigurationTelemetry te
...
-            ProbeFile = config.WithKeys(ConfigurationKeys.Debugger.DynamicInstrumentationProbeFile).AsString() ?? string.Empty;
+            // This is an internal test harness flag, not a registry-backed tracer configuration.
+#pragma warning disable RS0030
+            SnapshotExplorationTestRootPath = Environment.GetEnvironmentVariable(SnapshotExplorationConstants.TestRootPathKey) ?? string.Empty;
+#pragma warning restore RS0030
+            // Snapshot exploration is an internal test mode for the test host only. Child processes may inherit the
+            // root path, but they should keep the normal debugger sinks/pollers/rate limits.
+            IsSnapshotExplorationTestEnabled =
+                !StringUtil.IsNullOrEmpty(SnapshotExplorationTestRootPath) &&
+                SnapshotExplorationConstants.IsRunningInTestHost();
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/DebuggerSettings.cs:167]: 80603ff517879032bcef0507c34f20f2fafea7cc

File: tracer/src/Datadog.Trace/Debugger/Expressions/ProbeExpressionsProcessor.cs [hunk 21]
```diff
@@ -9,6 +9,7 @@
 using Datadog.Trace.Debugger.Configurations.Models;
 using Datadog.Trace.Logging;
 using Datadog.Trace.Util.Json;
+using Datadog.Trace.Vendors.Newtonsoft.Json;
 using Datadog.Trace.Vendors.Serilog.Events;
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Expressions/ProbeExpressionsProcessor.cs:12]: 80603ff517879032bcef0507c34f20f2fafea7cc

File: tracer/src/Datadog.Trace/Debugger/DynamicInstrumentation.cs [hunk 22]
```diff
@@ -432,6 +416,28 @@ private static LineProbeResolveErrorDetails GetLineProbeResolveErrorDetails(Line
             return values is { Length: > 0 } ? string.Join(" | ", values) : null;
         }
 
+        private void SetRateLimit(ProbeDefinition probe)
+        {
+            if (_settings.IsSnapshotExplorationTestEnabled)
+            {
+                ProbeRateLimiter.Instance.TryAddSampler(probe.Id, NopAdaptiveSampler.Instance);
+                return;
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/DynamicInstrumentation.cs:430]: 029ce4d68765cb776d7a3a6b97dbfab5ff558dc1

File: tracer/src/Datadog.Trace/Debugger/ProbeStatuses/SnapshotExplorationProbeStatusPoller.cs [hunk 23]
```diff
@@ -0,0 +1,322 @@
...
+            {
+                if (_isDisposed)
+                {
+                    return;
+                }
+
+                _probes.UnionWith(newProbes);
+            }
+
+            EmitKnownStatuses(newProbes);
```
[the remark this answers is not in the document]
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/ProbeStatuses/SnapshotExplorationProbeStatusPoller.cs:74]: 5b5a6c90510203a1f89a3f855bfc7fc4bf85934e

**Review by jpbempel — APPROVED** [review]

File: tracer/build/_build/Build.ExplorationTests.cs [hunk 24]
```diff
@@ -175,6 +175,45 @@ Target RunExplorationTests
...
         ;
 
+    Target SetUpSnapshotExplorationTests
+        => _ => _
+               .Description("Sets up the Snapshot Exploration Test")
+               .Requires(() => ExplorationTestUseCase)
+               .After(Clean, BuildTracerHome)
+               .Executes(() =>
+                {
+                    if (ExplorationTestUseCase != global::ExplorationTestUseCase.Debugger)
```
Comment (andrewlock) [inline tracer/build/_build/Build.ExplorationTests.cs:185]: It seems weird that you have to specify a `ExplorationTestUseCase`, but then this just bails out if it's not set to debugger? 🤔 Should we just make these debugger-specific anyway (they seem to be debugger specific already anyway, because of the whole "snapshot" part?)

File: tracer/build/_build/Build.ExplorationTests.cs [hunk 25]
```diff
@@ -175,6 +175,45 @@ Target RunExplorationTests
                 })
         ;
 
+    Target SetUpSnapshotExplorationTests
```
Comment (andrewlock) [inline tracer/build/_build/Build.ExplorationTests.cs:178]: Are these exploration tests going to _replace_ our existing exploration tests? I ask, because they currently flake a lot, they're fixed in time to (old) versions of libraries, and I'm not aware of them ever identifying a "real" issue... so should we just replace those old versions?

File: tracer/build/_build/Build.SnapshotExplorationTest.cs
Comment (andrewlock) [inline tracer/build/_build/Build.SnapshotExplorationTest.cs:1]: nit: Instead of making this a partial class on `Build`, move it into its own sub-folder, and pass in the variables you need...

The problem with the current approach is that all of these "private" methods are available in every single Nuke target, which is confusing when working on those, and breaks encapsulation

File: tracer/src/Datadog.Trace/Debugger/DebuggerFactory.cs [hunk 26]
```diff
@@ -115,6 +123,11 @@ private static DiagnosticsUploader CreateDiagnosticsUploader(IDiscoveryService d
 
     internal static IDebuggerUploader CreateSymbolsUploader(IDiscoveryService discoveryService, Func<string> serviceNameProvider, TracerSettings tracerSettings, DebuggerSettings settings, IGitMetadataTagsProvider gitMetadataTagsProvider)
     {
+        if (settings.IsSnapshotExplorationTestEnabled)
+        {
+            return NoOpSymbolUploader.Instance;
+        }
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/DebuggerFactory.cs:129]: nit: Is it necessary to add all these special cases in? If we can just disable this functionality using env variables externally, that seems preferable to bloating the size of production code and adding complexity?

File: tracer/src/Datadog.Trace/Debugger/DebuggerManager.cs [hunk 27]
```diff
@@ -216,7 +216,30 @@ private void EnsureSnapshotPipelineConfigured(DebuggerSettings settings)
...
-            return UpdateProductsState(tracerSettings, newDebuggerSettings ?? DebuggerSettings);
+            var settings = newDebuggerSettings ?? DebuggerSettings;
+
+            // Snapshot exploration test runs take a different initialization path (no agent
+            // discovery, no-op symbols/probe-status pollers, sink writes to CSV).
+            // Route here so that Instrumentation.cs stays free of debugger test plumbing.
+            // This intentionally runs and completes synchronously (returning an already-completed
+            // task): the test host must not start executing probed code until initialization and
+            // probe installation have finished. This path only runs inside a vstest test host.
+            if (settings.IsSnapshotExplorationTestEnabled)
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/DebuggerManager.cs:227]: It seems like we're adding a _lot_ of extra complexity here just to configure all this stuff for testing. Is there a good reason why we can't send all these snapshots to the test agent as usual, and just have the test agent dump them instead?

File: tracer/src/Datadog.Trace/Debugger/DebuggerSettings.cs [hunk 28]
```diff
@@ -151,9 +152,19 @@ public DebuggerSettings(IConfigurationSource? source, IConfigurationTelemetry te
                                          .AsInt32(DefaultCodeOriginExitSpanFrames, frames => frames > 0)
                                          .Value;
 
+            ProbeFile = config.WithKeys(ConfigurationKeys.Debugger.DynamicInstrumentationProbeFile).AsString() ?? string.Empty;
+
             SymbolDatabaseCompressionEnabled = config.WithKeys(ConfigurationKeys.Debugger.SymbolDatabaseCompressionEnabled).AsBool(true);
 
-            ProbeFile = config.WithKeys(ConfigurationKeys.Debugger.DynamicInstrumentationProbeFile).AsString() ?? string.Empty;
+            // This is an internal test harness flag, not a registry-backed tracer configuration.
+#pragma warning disable RS0030
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/DebuggerSettings.cs:160]: You must not disable these. You must add new settings to the config registry

File: tracer/src/Datadog.Trace/Debugger/SnapshotExplorationConstants.cs [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+    /// child processes spawned by tests.
+    /// </summary>
+    public const string TestHostName = "testhost";
+
+    /// <summary>
+    /// File pattern for the native CLR profiler log. The exploration test
+    /// setup tails these files to detect when the native side has accepted
+    /// the probe configuration.
+    /// </summary>
+    public const string NativeLogFilePattern = "dotnet-tracer-native-*.log";
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/SnapshotExplorationConstants.cs:37]: This feels like a _really_ nasty smell... this sort of thing _shouldn't_ be in the Datadog.Trace binary we ship IMO 😟 This should all be external...

File: tracer/test/Datadog.Trace.Tests/Debugger/DynamicInstrumentationTests.cs [hunk 30]
```diff
@@ -36,6 +36,7 @@
 
 namespace Datadog.Trace.Tests.Debugger;
 
+[Collection(nameof(EnvironmentVariablesTestCollection))]
```
Comment (andrewlock) [inline tracer/test/Datadog.Trace.Tests/Debugger/DynamicInstrumentationTests.cs:39]: This is a massive smell 🙁 We are trying to _remove_ these. I don't see a reason we can't just read this from configuration like "normal" instead? 🤔

File: tracer/test/Datadog.Trace.Tests/Telemetry/ConfigurationTests.cs [hunk 31]
```diff
@@ -61,6 +61,7 @@ public class ConfigurationTests
                 "_DD_ROOT_DOTNET_SESSION_ID",
                 "DD_INTERNAL_TRACE_NATIVE_ENGINE_PATH",
                 "DD_INTERNAL_PROFILING_NATIVE_ENGINE_PATH",
+                "DD_INTERNAL_SNAPSHOT_EXPLORATION_TEST_ROOT_PATH",
```
Comment (andrewlock) [inline tracer/test/Datadog.Trace.Tests/Telemetry/ConfigurationTests.cs:64]: You can't add more here 😉 

```suggestion
```

**Review by andrewlock — COMMENTED** [review]
I have a few concerns with the huge extra complexity this seems to add to production code... It feels like the vast majority of this _shouldn't_ be in Datadog.Trace. I would recommend considering if there's any way to get the same behaviour using a test agent that dumps things to file, if that's what you need. Ultimately it's up to you of course.

RE the build changes, I think we can simplify it a little, but mostly just move it out of being a Nuke partial `Build` class. Also, can we retire the old exploration tests now (as they're flaky and add little value IMO?) Separate PR if so.

Finally, the config needs to be registered in the config registry, we mustn't use `#pragma` etc to ignore the errors.

File: tracer/build/_build/Build.ExplorationTests.cs (hunk 25 above)
Comment (dudikeleti) [reply tracer/build/_build/Build.ExplorationTests.cs:178]: At this point, they're indeed a bit less necessary (unless we add new IL rewriting capabilities). Early on, they helped uncover issues that we probably wouldn't have caught ourselves until they surfaced at a customer's site.
That said, you're right that we could probably replace them. The downside is figuring out when to run them. For the existing exploration tests it's easy, we should run them whenever there are changes in the CLR profiler files. But those new tests are end-to-end tests. 🤔 
In any case, I think we'll need to clean this up. We could perhaps keep only the new tests, with a few different levels of test granularity.

File: tracer/build/_build/Build.ExplorationTests.cs (hunk 24 above)
Comment (dudikeleti) [reply tracer/build/_build/Build.ExplorationTests.cs:185]: Correct. I'll fix it.

File: tracer/src/Datadog.Trace/Debugger/SnapshotExplorationConstants.cs (hunk 29 above)
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/SnapshotExplorationConstants.cs:37]: ok lets talk offline

File: tracer/src/Datadog.Trace/Debugger/DebuggerManager.cs (hunk 27 above)
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/DebuggerManager.cs:227]: Unfortunately, this would not be enough.

File: tracer/src/Datadog.Trace/Debugger/DebuggerFactory.cs (hunk 26 above)
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/DebuggerFactory.cs:129]: I'm not sure I follow. I'll catch you offline to discuss it.
